src/net/socket_state.rs

```diff
@@ -40,30 +42,63 @@ impl SocketState {
     }
 
     /// This will initialize the seq number, ack number and give back the raw data of the packet with the updated information.
-    pub fn pre_process_packet(&mut self, packet: Packet) -> Result<(SocketAddr, Vec<u8>)> {
-        let connection = self.create_connection_if_not_exists(&packet.addr)?;
-        // queue new packet
-        if let Ok(mut lock) = connection.write() {
-            let seq_num = lock.seq_num;
-            lock.waiting_packets.enqueue(seq_num, packet.clone());
+    pub fn pre_process_packet(&mut self, packet: Packet, config: &NetworkConfig) ->  Result<(SocketAddr, PacketData)>  {
+
+        if packet.payload().len() > config.max_packet_size {
+            error!("Packet too large: Attempting to send {}, max={}", packet.payload().len(), config.max_packet_size);
+            return Err(NetworkError::ExceededMaxPacketSize.into());
         }
 
-        let raw_packet: RawPacket;
-        // initialize packet data, seq, acked_seq etc.
-        let mut l = connection
+        let connection = self.create_connection_if_not_exists(&packet.addr)?;
+
+        let mut lock = connection
             .write()
             .map_err(|_| NetworkError::AddConnectionToManagerFailed)?;
 
-        raw_packet = RawPacket::new(
-            l.seq_num,
-            &packet,
-            l.their_acks.last_seq,
-            l.their_acks.field,
-        );
-        // increase sequence number
-        l.seq_num = l.seq_num.wrapping_add(1);
-        let buffer = serialize(&raw_packet)?;
-        Ok((packet.addr, buffer))
+        let connection_seq = lock.seq_num;
+        // queue new packet
+        lock.waiting_packets.enqueue(connection_seq, packet.clone());
```

Comment:
For now, it does not really matter much though because the packet is not changed after it is enqueued waiting for acknowledgment. The packet is content will be ready before it is enqueued. What we can do to avoid change is to make the fields of Packet nonpublic and only readable. So once a Packet is defined you can change it anymore.

Reply:
I have changed all fields of the packet to private and made getters for them so they can not be changed once declared.

Reply:
Okay, thats fine, I was just thinking that this would be a good use case for a ref counter.